File: app/src/main/java/com/dougkeen/bart/activities/DeparturesViewModel.java
package com.dougkeen.bart.activities;

import android.content.Context;

import androidx.lifecycle.ViewModel;

import com.dougkeen.bart.backend.RouteDepartureProjection;
import com.dougkeen.bart.backend.TransitFeedSnapshot;
import com.dougkeen.bart.backend.TransitProjectionListener;
import com.dougkeen.bart.backend.TransitRepository;
import com.dougkeen.bart.model.Departure;
import com.dougkeen.bart.model.RealTimeDepartures;
import com.dougkeen.bart.model.StationPair;
import com.dougkeen.bart.model.SystemTimeSource;
import com.dougkeen.bart.model.TimeSource;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/** Screen-owned realtime state retained across departures-screen recreation. */
public final class DeparturesViewModel extends ViewModel {
    public enum Status {
        LOADING,
        CONTENT,
        EMPTY,
        ERROR
    }

    public interface Listener {
        void onStateChanged(State state);
    }

    public static final class State {
        private final Status status;
        private final List<Departure> departures;
        private final Exception error;

        private State(Status status, List<Departure> departures, Exception error) {
            this.status = status;
            this.departures = departures;
            this.error = error;
        }

        public Status getStatus() {
            return status;
        }

        public List<Departure> getDepartures() {
            return departures;
        }

        public Exception getError() {
            return error;
        }

        private static State loading() {
            return new State(Status.LOADING, Collections.emptyList(), null);
        }

        private static State content(List<Departure> departures) {
            return new State(Status.CONTENT, departures, null);
        }

        private static State empty() {
            return new State(Status.EMPTY, Collections.emptyList(), null);
        }

        private static State error(Exception exception, List<Departure> departures) {
            return new State(Status.ERROR, departures, exception);
        }
    }

    private final TimeSource timeSource;
    private List<Departure> departures = Collections.emptyList();
    private State state = State.loading();
    private Listener listener;
    private TransitRepository.Subscription subscription;

    public DeparturesViewModel() {
        this(SystemTimeSource.INSTANCE);
    }

    public DeparturesViewModel(TimeSource timeSource) {
        this.timeSource = timeSource;
    }

    public synchronized State getState() {
        return state;
    }

    public void start(TransitRepository repository, Context context,
                      StationPair stationPair, Listener listener) {
        State currentState;
        synchronized (this) {
            stopSubscriptionLocked();
            this.listener = listener;
            currentState = state;
            if (stationPair != null) {
                subscription = repository.subscribe(
                        new RouteDepartureProjection(stationPair, context),
                        new TransitProjectionListener<RealTimeDepartures>() {
                            @Override
                            public void onData(RealTimeDepartures result,
                                               TransitFeedSnapshot snapshot) {
                                updateFromFeed(result.getDepartures());
                            }

                            @Override
                            public void onError(Exception exception,
                                                TransitFeedSnapshot snapshot) {
                                updateError(exception);
                            }
                        });
            }
        }
        if (listener != null) {
            listener.onStateChanged(currentState);
        }
    }

    public synchronized void stop() {
        stopSubscriptionLocked();
        listener = null;
    }

    public synchronized List<Departure> replace(List<Departure> incoming) {
        departures = immutableCopy(Departure.replaceFeed(departures, incoming, timeSource));
        state = departures.isEmpty() ? State.empty() : State.content(departures);
        return departures;
    }

    public synchronized List<Departure> clear() {
        departures = Collections.emptyList();
        state = State.empty();
        return departures;
    }

    public synchronized List<Departure> getDepartures() {
        return departures;
    }

    private static List<Departure> immutableCopy(List<Departure> values) {
        return Collections.unmodifiableList(new ArrayList<>(values));
    }

    @Override
    protected void onCleared() {
        stop();
        synchronized (this) {
            departures = Collections.emptyList();
            state = State.loading();
        }
        super.onCleared();
    }

    private void updateFromFeed(List<Departure> incoming) {
        Listener currentListener;
        State nextState;
        synchronized (this) {
            departures = immutableCopy(Departure.replaceFeed(departures, incoming, timeSource));
            state = departures.isEmpty() ? State.empty() : State.content(departures);
            nextState = state;
            currentListener = listener;
        }
        if (currentListener != null) {
            currentListener.onStateChanged(nextState);
        }
    }

    private void updateError(Exception exception) {
        Listener currentListener;
        State nextState;
        synchronized (this) {
            state = State.error(exception, departures);
            nextState = state;
            currentListener = listener;
        }
        if (currentListener != null) {
            currentListener.onStateChanged(nextState);
        }
    }

    private void stopSubscriptionLocked() {
        if (subscription != null) {
            subscription.close();
            subscription = null;
        }
    }
}

File: app/src/main/kotlin/com/dougkeen/bart/activities/DeparturesViewModel.kt
package com.dougkeen.bart.activities

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dougkeen.bart.backend.RouteDepartureProjection
import com.dougkeen.bart.backend.TransitRepository
import com.dougkeen.bart.model.Departure
import com.dougkeen.bart.model.StationPair
import com.dougkeen.bart.model.SystemTimeSource
import com.dougkeen.bart.model.TimeSource
import com.dougkeen.bart.transit.gtfs.BartGtfsNetwork
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import java.util.ArrayList
import java.util.Collections
import java.util.function.Supplier

/** Owns departures-screen state and its shared-feed projection. */
class DeparturesViewModel @JvmOverloads constructor(
    private val timeSource: TimeSource = SystemTimeSource,
) : ViewModel() {
    enum class Status {
        LOADING,
        CONTENT,
        EMPTY,
        ERROR,
    }

    data class State(
        val status: Status,
        val departures: List<Departure>,
        val error: Exception?,
    ) {
        companion object {
            fun loading() = State(Status.LOADING, emptyList(), null)
            fun content(departures: List<Departure>) =
                State(Status.CONTENT, departures, null)
            fun empty() = State(Status.EMPTY, emptyList(), null)
            fun error(exception: Exception, departures: List<Departure>) =
                State(Status.ERROR, departures, exception)
        }
    }

    private val _uiState = MutableStateFlow(State.loading())
    val uiState: StateFlow<State> = _uiState.asStateFlow()

    private var departures: List<Departure> = emptyList()
    private var collectionJob: Job? = null

    /** Configures the query; the ViewModel owns collection until it is cleared. */
    fun setQuery(
        repository: TransitRepository,
        networkSupplier: Supplier<BartGtfsNetwork>,
        stationPair: StationPair?,
    ) {
        collectionJob?.cancel()
        collectionJob = null
        departures = emptyList()
        _uiState.value = State.loading()
        if (stationPair == null) {
            return
        }

        collectionJob = viewModelScope.launch {
            val projection = RouteDepartureProjection(stationPair, networkSupplier)
            repository.projectedState(projection::project, projection::areEquivalent)
                .collectLatest { result ->
                    result.exceptionOrNull()?.let { exception ->
                        updateError(asException(exception))
                    } ?: result.getOrNull()?.let { departures ->
                        updateFromFeed(departures.getDepartures())
                    }
                }
        }
    }

    @Synchronized
    fun replace(incoming: List<Departure>): List<Departure> {
        departures = immutableCopy(Departure.replaceFeed(departures, incoming, timeSource))
        _uiState.value = if (departures.isEmpty()) {
            State.empty()
        } else {
            State.content(departures)
        }
        return departures
    }

    @Synchronized
    fun clear(): List<Departure> {
        departures = emptyList()
        _uiState.value = State.empty()
        return departures
    }

    fun getState(): State = uiState.value

    @Synchronized
    fun getDepartures(): List<Departure> = departures

    @Synchronized
    private fun updateFromFeed(incoming: List<Departure>) {
        replace(incoming)
    }

    @Synchronized
    private fun updateError(exception: Exception) {
        _uiState.value = State.error(exception, departures)
    }

    private fun immutableCopy(values: List<Departure>): List<Departure> =
        Collections.unmodifiableList(ArrayList(values))

    private fun asException(error: Throwable): Exception =
        error as? Exception ?: RuntimeException(error)
}
